### PR TITLE
Adds a special handling for DataVolumes in WaitForFirstConsumer state.

### PR DESCRIPTION
**Pull request by brybacki**
**What this PR does / why we need it**:

Adds a special handling for DataVolumes in WaitForFirstConsumer state.

In order to support CDI's delayed binding mode the kubevirt's VMI now has a special handling for a VMI that depends on DataVolumes which are in Pending/WaitForFirstConsumer status. In this case, the controller should reflect that the VMI is in a special "Provisioning" state. To move out of this state, the Controller creates a virt-launcher Pod which does not run the real VM workload. This causes all the PVCs associated with the Pending DataVolume to become Bound. Future iterations of the VMI control loop will find no DataVolumes that are Pending and the VMI status can move to running and the VM can be launched as normal.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Adds a special handling for DataVolumes in WaitForFirstConsumer state to support CDI's delayed binding mode.
```

### CODE REVIEW

File: pkg/virt-controller/watch/vmi.go
Comment (davidvossel) [inline pkg/virt-controller/watch/vmi.go:622]: it might be wise for us to put a special label or annotation on these temp pods so we can differentiate them and filter them differently from the vmi pods.

**Review by dhiller — COMMENTED** [review]
In general the PR looks fine. Maybe you might consider to not use the abbreviation `wffc` for future readers to be more clear?

File: staging/src/kubevirt.io/client-go/api/v1/types.go
Comment (brybacki) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:423]: Yes, but to do this I would need to differentiate between vmPod and tempPod (as you note in next comment). So let me try to do this.

**Comment by brybacki** [comment]
> In general the PR looks fine. Maybe you might consider to not use the abbreviation `wffc` for future readers to be more clear?

Good catch, I forgot that what is easier to write might be harder to read/understand.

File: pkg/virt-controller/watch/vmi.go
Comment (davidvossel) [inline pkg/virt-controller/watch/vmi.go:604]: I thought about this a bit over the weekend. One thing that this PR doesn't handle is a situation where the tempPod fails.  It's techinically possible for a pod to be scheduled to a node, and then fail to launch on that node for reasons unrelated to the workload. 

Because of this, a tempPod could be created and never technically serve it's purpose, then I believe the VMI will get stuck in pending forever.

To fix this edge case, we could make the tempPod a k8s Job instead, which guarantees that the pod will be retried until it eventually succeeds somewhere. You can look at `pkg/virt-operator/kubevirt.go` for an example of how we used a job in another controller.

**Review by davidvossel — COMMENTED** [review]
added a comment about using a k8s Job instead of pod directly.

File: staging/src/kubevirt.io/client-go/api/v1/types.go
Comment (brybacki) [reply staging/src/kubevirt.io/client-go/api/v1/types.go]: Done

File: pkg/virt-controller/watch/vmi.go
Comment (brybacki) [reply pkg/virt-controller/watch/vmi.go:627]: Added annotation

File: pkg/virt-controller/watch/vmi.go
Comment (brybacki) [reply pkg/virt-controller/watch/vmi.go:604]: I am looking at this right now, in the meantime a rebased/fresh version was pushed to check if there are any failing tests due to new featureflag being set.

Comment (brybacki) [reply pkg/virt-controller/watch/vmi.go:604]: good, so I can easily wrap a pod in a job, but as I did not use jobs yet I have a few questions
- do you think we need some special settings for backoff and restart 
- will the pod nicely be deleted after job finishes or do I need to find and kill a job or a pod later?

**Comment by brybacki** [comment]
on my env there is nothing after running `make generate` and `make generate-verify` passes, but I have seen a few issues mentioning this already

File: tests/datavolume_test.go
Comment (aglitke) [inline tests/datavolume_test.go:206]: You need to remove this Focus test before merge.

**Comment by dhiller** [comment]
> on my env there is nothing after running `make generate` and `make generate-verify` passes, but I have seen a few issues mentioning this already

There's a PR in the works to fix the cause (swagger generation inconsistency): https://github.com/kubevirt/kubevirt/pull/4154

File: tests/datavolume_test.go
Comment (brybacki) [reply tests/datavolume_test.go]: right, done

File: pkg/virt-controller/watch/vmi.go
Comment (brybacki) [reply pkg/virt-controller/watch/vmi.go:604]: ok, so after cleaning up other problems, I took a try at this, and this looks like a doable and not so hard task, but we for sure can not make it today/tomorrow, there are some rules/access rights to change, tests to update and small details to analyze. For example I get this when trying to create a job:
```
obs.batch is forbidden: User "system:serviceaccount:kubevirt:kubevirt-controller" cannot create resource "jobs" in API group "batch" in the namespace "kubevirt-test-default"
```

So @davidvossel, @aglitke, I am asking, what is the risk for leaving it as is right now as the `pod` (this new feature that fixes some other problems), and doing a conversion to `job` after merge as a new follow-up task.

**Comment by brybacki** [comment]
I needed to push small 'hack; for cdi to enable more logging to check what is wrong with importer in 3189 tests, so this should not be merged right now,

File: pkg/virt-controller/watch/vmi.go
Comment (davidvossel) [inline pkg/virt-controller/watch/vmi.go:352]: These if statements short circuit the final `else` below, which both sets the vmi phase to `pending` and reports any potential sync errors as a condition.

Regardless of the wffc logic, we still need to make sure sync errors are visible as a condition.

File: pkg/virt-controller/watch/vmi.go
Comment (davidvossel) [inline pkg/virt-controller/watch/vmi.go:345]: is there a reason we should be removing the condition if it exists only to re-add it again?

File: pkg/virt-controller/watch/vmi.go
Comment (davidvossel) [inline pkg/virt-controller/watch/vmi.go:351]: i'm not sure we're guaranteed the provisioning condition will get removed as it is now.  Do we have any guarantee that the vmi won't transition from pending->scheduled until this condition is removed? if that happens, then the pending condition will be around indefinitely.

File: pkg/virt-controller/watch/vmi.go
Comment (davidvossel) [reply pkg/virt-controller/watch/vmi.go:604]: > So @davidvossel, @aglitke, I am asking, what is the risk for leaving it as is right now as the pod (this new feature that fixes some other problems), and doing a conversion to job after merge as a new follow-up task.

The alternative to the job logic is to tie the lifecycle of the VMI to the temp pod... so if the temp pod fails for any reason, immediately fail the VMI just like we would the real pod. 

The situation we don't want to be in is having a VMI stuck in pending indefinitely due to a failed temp pod that will never get retried and will never cause the VMI to move out of pending.

File: pkg/virt-controller/watch/vmi.go
Comment (davidvossel) [inline pkg/virt-controller/watch/vmi.go:626]: when we talk about a vmi being stuck indefinitely, this is specifically the place that occurs. We wait forever for the isWatiForFirstConsumer to become false without checking the status of that Pod (other than knowing it exists)

File: pkg/virt-controller/watch/vmi.go
Comment (davidvossel) [inline pkg/virt-controller/watch/vmi.go:628]: we need to have a delete expectation set here.

**Review by davidvossel — COMMENTED** [review]
looks pretty close. I made a few more comments inline 

Functionally, the only big change here is we need to make sure we're either using a job to execute the temp job, or tying the temp job's lifecycle to the VMI's lifecycle.

So, with a job, if the temp pod fails it will be retried.

And the alternative of tying the VMI to the lifecycle of the temp pod results in the VMI failing if the temp pod fails.

I think the simplest thing to do here is stick with the latter approach, of simply failing the VMI if the temp pod fails. (sorry i went back and forth on this, i just needed some time to think it through further) That's essentially the same behavior we'd have if the real VMI pod itself failed anyway. We'd be depending on the VM controller to decide whether to relaunch the VMI or not.

File: pkg/virt-controller/watch/vmi.go
Comment (brybacki) [reply pkg/virt-controller/watch/vmi.go:345]: no reason, removed

File: pkg/virt-controller/watch/vmi.go
Comment (brybacki) [reply pkg/virt-controller/watch/vmi.go:628]: ok

File: pkg/virt-controller/watch/vmi.go
Comment (brybacki) [reply pkg/virt-controller/watch/vmi.go:351]: now I think its place is in the `	case vmi.IsScheduling():` branch of the switch, then we are sure provisioning has ended.

File: pkg/virt-controller/watch/vmi.go
Comment (brybacki) [inline pkg/virt-controller/watch/vmi.go:521]: @davidvossel, @mhenriks Here I check for failing temp Pod. Do you think there are other scenarios than simple Failed state, or Unschedulable pod (I think I need to check for this unschedulable state)? An what else I should check here?

Basically if a pod is successfully scheduled then the PVCs should be bound. I am trying to think about other pathological cases to cover here, so we do not miss any problems.

Comment (brybacki) [reply pkg/virt-controller/watch/vmi.go:521]: I have now added check for pod condition PodScheduled with Reason == k8sv1.PodReasonUnschedulable, and mark VMI as failed. 
Now when I test this I can see that VMI fails and is retried by controller. Is it good behavior? VM shows this:
```
Events:
  Type    Reason                      Age                   From                       Message
  ----    ------                      ----                  ----                       -------
  Normal  SuccessfulDataVolumeCreate  23m                   virtualmachine-controller  Created DataVolume alpine-dv
  Normal  SuccessfulDelete            21m                   virtualmachine-controller  Stopped the virtual machine by deleting the virtual machine instance 871a8973-0f2e-4744-852f-cfeb85c36fc3
  Normal  SuccessfulDelete            21m                   virtualmachine-controller  Stopped the virtual machine by deleting the virtual machine instance d62d5c34-5dc6-46df-924d-566e0ab20eb3
  Normal  SuccessfulDelete            21m                   virtualmachine-controller  Stopped the virtual machine by deleting the virtual machine instance 967b46fa-0022-4d1f-b431-a6304fc96ab2
  Normal  SuccessfulDelete            21m                   virtualmachine-controller  Stopped the virtual machine by deleting the virtual machine instance 11940934-ae44-4a36-8276-1f10ab862f41
  Normal  SuccessfulDelete            21m                   virtualmachine-controller  Stopped the virtual machine by deleting the virtual machine instance 0ff178fd-e979-402f-84bb-6b8539f8ae4f
  Normal  SuccessfulDelete            21m                   virtualmachine-controller  Stopped the virtual machine by deleting the virtual machine instance 5e0d40d9-b4fe-45fe-9523-b97558afbd72
  Normal  SuccessfulDelete            21m                   virtualmachine-controller  Stopped the virtual machine by deleting the virtual machine instance a7e706e8-559f-4a15-8dcc-0fe876bfe832
  Normal  SuccessfulDelete            21m                   virtualmachine-controller  Stopped the virtual machine by deleting the virtual machine instance 4e79a8c0-b699-4298-a140-d13b9e3e3323
  Normal  SuccessfulDelete            21m                   virtualmachine-controller  Stopped the virtual machine by deleting the virtual machine instance 34c3dcb8-8e63-4135-a860-89974c046ce5
  Normal  SuccessfulCreate            21m (x12 over 21m)    virtualmachine-controller  Started the virtual machine by creating the new virtual machine instance vm-alpine-datavolume
  Normal  SuccessfulDelete            3m6s (x564 over 21m)  virtualmachine-controller  (combined from similar events): Stopped the virtual machine by deleting the virtual machine instance 5e424c0f-3176-4a7a-9394-3fa6d4f5714e

```

Comment (mhenriks) [reply pkg/virt-controller/watch/vmi.go]: I think the pod will just be pending forever when can't schedule so not sure you can deal with that without maybe looking at events.  Should be easy enough to test?

File: pkg/virt-controller/watch/vmi.go
Comment (davidvossel) [inline pkg/virt-controller/watch/vmi.go:360]: We should maintain the same behavior for this temporary pod as we have for the normal pod that actually hosts the VMI. The PodReasonUnschedulable is a transient condition. It can clear up if more resources are added to the cluster

```
        // PodReasonUnschedulable reason in PodScheduled PodCondition means that the scheduler
        // can't schedule the pod right now, for example due to insufficient resources in the cluster.
```

Right now, VMIs don't fail if this condition is encountered with the pod that hosts the qemu process. The behavior for the temporary pod shoudl be the same. 

To maintain the same behavior as it is today, we'd just fail the VMI if the temp pod is going down or is down.

File: pkg/virt-controller/watch/vmi.go
Comment (davidvossel) [inline pkg/virt-controller/watch/vmi.go:346]: wouldn't this `Provisioning` condition need to be set regardless if the temp pod exists or not? Otherwise there's no guarantee that the provisioning condition will get set. There's a race between setting this condition and creating the temp pod.

For instance, if update status fails to post the update to the VMI's status, then it's possible the next iteration that the temp pod exists and the provisioning condition is never set.

File: pkg/virt-controller/watch/vmi.go
Comment (davidvossel) [reply pkg/virt-controller/watch/vmi.go]: looks like you might have resolved this?

Comment (brybacki) [reply pkg/virt-controller/watch/vmi.go]: looks like it is by design, ` runStrategy: "RerunOnFailure` vs `runStrategy: Manual`

File: pkg/virt-controller/watch/vmi.go
Comment (brybacki) [reply pkg/virt-controller/watch/vmi.go:346]: you are right, it is simple, if WaitForFirstConsumer then controler needs to run special logic to force provisioning (provisioning condition set), then vmi transitions to Scheduling phase and controller can remove the condition!

File: pkg/virt-controller/watch/vmi.go
Comment (brybacki) [reply pkg/virt-controller/watch/vmi.go:360]: Ok if that is the case, then I am updating it - removing the check for PodReasonUnschedulable.

File: tests/datavolume_test.go
Comment (davidvossel) [inline tests/datavolume_test.go:84]: i think you'll want to capture what the cdi config has set before the test, and then only enable/disable based on whether cdi was configured to HonorWaitForFirstConsume. it's possible that setting might be configured as part of the environment setup before the test suite runs.

File: tests/datavolume_test.go
Comment (davidvossel) [inline tests/datavolume_test.go:141]: we don't need this anymore, right? This is what the vmi.go logic is doing with the temp pod?

**Review by davidvossel — COMMENTED** [review]
looking good, i just had some questions around the functional tests.  It's unclear to me if we're exercising the new behavior.

File: tests/datavolume_test.go
Comment (brybacki) [reply tests/datavolume_test.go:84]: I did try to configure it in `cluster-deploy.sh` script, but unfortunately there are destructive tests (probably this one https://github.com/kubevirt/kubevirt/blob/master/tests/operator_test.go#L1414 this one maybe?) redeploying CDI using api alphav1 and by side-effect resetting new config option. 
That is why after discovering this I have moved enable/disable to before/after test.

File: tests/datavolume_test.go
Comment (brybacki) [reply tests/datavolume_test.go:141]: this is used in test in places where test creates Datavolume and waits for it to start import, but... maybe that logic is not needed -- checking

Comment (brybacki) [reply tests/datavolume_test.go:141]: I thought that in this test it might be needed, because it needs to have a DV with bad url and then it tests the eventual consistency of DV. But I see that the test does not need to wait for dv to be ready.

**Comment by brybacki** [comment]
checking pull-kubevirt-check-tests-for-flakes again to see if the timeout problem persists, I suspect something might be overloaded? or maybe not

**Comment by brybacki** [comment]
@davidvossel What do you think about this PR? Is it ready to be merged?

File: tests/datavolume_test.go
Comment (davidvossel) [reply tests/datavolume_test.go:84]: the enable before/after is fine. 

My point was in the Before(), i'd rather see us capture the original contents on the ConfigMap before we alter it, then in the After() i'd like to see us restore the contents.

This ensures we preserve any featuregates that might have been configured as part of the test setup before the functional tests run. As it is now, the After() function will remove all feature gates.

**Review by davidvossel — COMMENTED** [review]
hey, sorry for the delayed response. The last thing on my list here is simply to ensure we don't accidentally stop on the cdi configMap during the func tests. I added some more details about my thoughts in line there.

**Comment by brybacki** [comment]
updated the Before/After part of test and rebased the PR (I have squashed some commits so the rebase was easier)

**Comment by brybacki** [comment]
looks like automatic merge during rebase was not successful, need to check what is the problem

File: pkg/virt-controller/watch/vmi.go
Comment (davidvossel) [inline pkg/virt-controller/watch/vmi.go:660]: how that PVCs can be referenced instead of the datavolume, don't we need to check for wffc in the event the PVC is referenced?

https://github.com/kubevirt/kubevirt/pull/4277

**Review by davidvossel — COMMENTED** [review]
I think https://github.com/kubevirt/kubevirt/pull/4277 may have introduced some implications that impact this PR.

File: pkg/virt-controller/watch/vmi.go
Comment (brybacki) [reply pkg/virt-controller/watch/vmi.go:660]: Right, after merging #4277, kubevirt correctly works with WFFC DV, but it does not work if the PVC is used instead of a DV.
It can be handled, but then we would need to follow the owner and get the original DV (so basically do everything that was hidden in the new utility function Is cdiv1.IsPopulated). We are already working on this.

Comment (davidvossel) [reply pkg/virt-controller/watch/vmi.go:660]: @brybacki what should be the expectation for this PR. are we waiting for the cdi change to land so this pr can leverage it?

Comment (brybacki) [reply pkg/virt-controller/watch/vmi.go:660]: Yes: https://github.com/kubevirt/containerized-data-importer/pull/1442

Should I mark this as on hold until new cdi version can be used?

Comment (davidvossel) [reply pkg/virt-controller/watch/vmi.go:660]: yup, makes sense to me.

**Comment by brybacki** [comment]
/hold 
waits for https://github.com/kubevirt/kubevirt/pull/4434, then needs some updates to handle PVC

File: pkg/virt-controller/watch/vmi.go
Comment (brybacki) [reply pkg/virt-controller/watch/vmi.go:674]: updated, PVC's are no covered

**Comment by brybacki** [comment]
strange result for 1.19 fail, there is no test pvc, no logs for cdi-controller `unable to retrieve container logs for docker://4d22f354f721a5cd621784cfac35c5bd77af12708c61dcec147379651d5a337e`

/retest

File: pkg/virt-controller/watch/vmi.go
Comment (brybacki) [reply pkg/virt-controller/watch/vmi.go:674]: @davidvossel can you check again? new stuff is in last commit `check PVC if it waits for first consumer …`

File: tests/datavolume_test.go
Comment (davidvossel) [inline tests/datavolume_test.go:743]: is this function ever used?

File: tests/datavolume_test.go
Comment (davidvossel) [inline tests/datavolume_test.go:111]: do we have a functional test that exercises the flow of when a VMI references a DV by PVC when wffc is used?

Comment (brybacki) [reply tests/datavolume_test.go:111]: looks like there is none, and this one can be easily added, buy converting current test to table, and changing VMI definition

File: tests/datavolume_test.go
Comment (brybacki) [reply tests/datavolume_test.go:743]: no longer used, I'll just remove it

File: tests/datavolume_test.go
Comment (brybacki) [reply tests/datavolume_test.go:111]: or  better - I can just enable WFFC for the 5252 test

**Review by davidvossel — APPROVED** [review]
/approve
/lgtm

**Comment by brybacki** [comment]
need to check if the test has problems or the implementation is bad

**Comment by brybacki** [comment]
interesting, the new tests do not fail (last 2 runs)
/retest

**Comment by brybacki** [comment]
still experimenting
/test pull-kubevirt-e2e-k8s-1.19
/test pull-kubevirt-e2e-k8s-1.18

**Comment by brybacki** [comment]
this failed on some unrelated timeout pull-kubevirt-e2e-k8s-1.17
this failed on cluster preparation! pull-kubevirt-e2e-k8s-cnao-1.17
this also unrelated pull-kubevirt-check-tests-for-flakes
7h run pull-kubevirt-e2e-k8s-1.19 
cluster init pull-kubevirt-e2e-k8s-1.18

so many fails, none connected to my changes, I'll revert my debug logs, and try to get all green

**Comment by brybacki** [comment]
finally some meaningfull logs -> https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4025/pull-kubevirt-check-tests-for-flakes/1329412576387272704

found inconsistency in implementation of helper function in CDI: IsWaitForFirstConsumerBeforePopulating

**Comment by brybacki** [comment]
this is a fix for inconsistent behavior: https://github.com/kubevirt/containerized-data-importer/pull/1499

so I'll need to use newer CDI after 1499 is merged

**Comment by davidvossel** [comment]
> this is a fix for inconsistent behavior: kubevirt/containerized-data-importer#1499 so I'll need to use newer CDI after 1499 is merged

@brybacki Any info on how this cdi PR fixes the functional test flakiness here?

Is it because the DataVolume can technically still briefly be in a WFFC state while the PVC has already moved to phase=Bound?

**Comment by brybacki** [comment]
> > this is a fix for inconsistent behavior: [kubevirt/containerized-data-importer#1499](https://github.com/kubevirt/containerized-data-importer/pull/1499) so I'll need to use newer CDI after 1499 is merged
> 
> @brybacki Any info on how this cdi PR fixes the functional test flakiness here?
> 
> Is it because the DataVolume can technically still briefly be in a WFFC state while the PVC has already moved to phase=Bound?

yes, 
And when looking at the DV in one place and at the PVC in the other - we got inconsistent view. I think it is better if we always check DV in this case, then we are 100% sure CDI importert/uploader/cloner already finished its job.

some more details (https://storage.googleapis.com/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4025/pull-kubevirt-check-tests-for-flakes/1329412576387272704/artifacts/k8s-reporter/pods/1_kubevirt_virt-controller-549977cd8c-xgqnx-virt-controller.log): 
1. the `sync` used info from PVC (through handleSyncDataVolumes/IsWaitForFirstConsumerBeforePopulating) and saw it is BOUND, already populated (no WFFC) so it `deleted temp Pod`, 
2. the `updateStatus` used info from DV wich was still in WFFC, but all was good in this execution, 
3. the next time controller run into `updateStatus`, it still saw DV in WFFC and `pod terminating`, and this is invalid state so it marked VMI as failed.

**Comment by brybacki** [comment]
so it is in fact not a test flakiness, but actual logical error (an egde case, depending on timing between events etc...)

**Comment by davidvossel** [comment]
> the next time controller run into updateStatus, it still saw DV in WFFC and pod terminating, and this is invalid state so it marked VMI as failed.

makes total sense. Great catch and great job powering through this feature. This is really tricky logic. You're definitely a pro now with the vmi controller :D

**Comment by brybacki** [comment]
Last push is a rebase to include changes from https://github.com/kubevirt/kubevirt/pull/4585, 
and 3 commits where squashed, no changes to code.

**Comment by rmohr** [comment]
CDI is updasted. Restoring Davids lgtm.

/lgtm

**Comment by rmohr** [comment]
/override pull-kubevirt-e2e-k8s-cnao-1.17 

This job seems to be gone.

**Pull request merged**
